### ISSUE TITLE
nn.Transformer gives different output in torch.no_grad() context

### ISSUE THREAD

**Issue by IndoorAdventurer** [pytorch/pytorch#128209]
### 🐛 Describe the bug

The output that `nn.Transformer` gives in `eval` mode with the same inputs, is different when the forward pass is performed in a `torch.no_grad()` context:
```python
# Defining the model and putting it in eval mode:
tf = torch.nn.Transformer(
    d_model=64,
    nhead=8,
    num_encoder_layers=1,
    num_decoder_layers=2,
    dim_feedforward=256,
    batch_first=True
)
tf.eval()


# Creating dummy input data:
batch_size = 8
src_inp = []
src_pmask = []
tgt_inp = []
tgt_pmask = []

for idx in range(batch_size):
    seq_len_src = torch.randint(5, 10, (1,))[0]
    src_inp.append(torch.rand((seq_len_src, 64)))
    src_pmask.append(torch.zeros((seq_len_src)))

    seq_len_tgt = torch.randint(5, 10, (1,))[0]
    tgt_inp.append(torch.rand((seq_len_tgt, 64)))
    tgt_pmask.append(torch.zeros((seq_len_tgt)))

src_inp = torch.nn.utils.rnn.pad_sequence(src_inp, True, -1)
src_pmask = torch.nn.utils.rnn.pad_sequence(src_pmask, True, float("-inf"))
tgt_inp = torch.nn.utils.rnn.pad_sequence(tgt_inp, True, -1)
tgt_pmask = torch.nn.utils.rnn.pad_sequence(tgt_pmask, True, float("-inf"))

# Running it with and without no_grad(), and comparing:
output1 = tf(
    src=src_inp,
    tgt=tgt_inp,
    tgt_mask=tf.generate_square_subsequent_mask(tgt_inp.size(1)),
    src_key_padding_mask=src_pmask,
    tgt_key_padding_mask=tgt_pmask,
    memory_key_padding_mask=src_pmask
)

with torch.no_grad():
    output2 = tf(
        src=src_inp,
        tgt=tgt_inp,
        tgt_mask=tf.generate_square_subsequent_mask(tgt_inp.size(1)),
        src_key_padding_mask=src_pmask,
        tgt_key_padding_mask=tgt_pmask,
        memory_key_padding_mask=src_pmask
    )

print(torch.all(output1 == output2))
```
The output I get, is: `[...]/python3.12/site-packages/torch/nn/modules/transformer.py:384: UserWarning: The PyTorch API of nested tensors is in prototype stage and will change in the near future. (Triggered internally at ../aten/src/ATen/NestedTensorImpl.cpp:177.)
  output = torch._nested_tensor_from_mask(output, src_key_padding_mask.logical_not(), mask_check=False)`, and `tensor(False)`.

The issue seems very similar to [this one, created in 2022](https://github.com/pytorch/pytorch/issues/86120), and likely has something to do with the fact that the model converts the encoder input into a nested tensor for some reason. When looking at the absolute difference between outputs `print(torch.abs(output1 - output2).mean())`, the difference seems failry minimal; in the order of 1e-7. Before, when I forgot to set a memory mask, the difference was much larger and substantially effected performance of my model during inference.

Kind regards,
Vincent

### Versions

Collecting environment information...
PyTorch version: 2.2.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Fedora Linux 39 (Workstation Edition) (x86_64)
GCC version: (GCC) 13.2.1 20240316 (Red Hat 13.2.1-7)
Clang version: Could not collect
CMake version: version 3.27.7
Libc version: glibc-2.38

Python version: 3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)] (64-bit runtime)
Python platform: Linux-6.7.10-200.fc39.x86_64-x86_64-with-glibc2.38
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 550.67
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i5-10300H CPU @ 2.50GHz
CPU family:                           6
Model:                                165
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             2
CPU(s) scaling MHz:                   85%
CPU max MHz:                          4500,0000
CPU min MHz:                          800,0000
BogoMIPS:                             4999,90
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            128 KiB (4 instances)
L1i cache:                            128 KiB (4 instances)
L2 cache:                             1 MiB (4 instances)
L3 cache:                             8 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==2.2.4
[pip3] torch==2.2.2
[pip3] torchaudio==2.2.2
[pip3] torchmetrics==1.4.0
[pip3] torchvision==0.17.2
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @ezyang @gchanan @zou3519 @kadeng @msaroufim

**Comment by IndoorAdventurer** [comment]
PS: just updated to 2.3.1, and it still occurs.

**Comment by Bekovmi** [comment]
model.eval() will notify all your layers that you are in eval mode, that way, batchnorm or dropout layers will work in eval mode instead of training mode.
torch.no_grad() impacts the autograd engine and deactivate it. It will reduce memory usage and speed up computations but you won’t be able to backprop (which you don’t want in an eval script).
